### ISSUE TITLE
DDIM Sampling

### ISSUE THREAD

**Issue by Saupatil07** [Janspiry/Palette-Image-to-Image-Diffusion-Models#63]
Hi, I wanted to know whether is it possible to use DDIM sampling for Palette instead of DDPM sampling as it takes lot of time. I tried using DDIM for SR3, but could not get good results. Any suggestion or help is appreciated.

**Comment by Janspiry** [comment]
Theoretically there is certainly no problem. But I haven't done any relevant research recently, so I'm sorry I may not have the relevant experience

**Comment by TMDTom** [comment]
> Hi, I wanted to know whether is it possible to use DDIM sampling for Palette instead of DDPM sampling as it takes lot of time. I tried using DDIM for SR3, but could not get good results. Any suggestion or help is appreciated.

You can find what you want int the StableDiffsion Code, there is a ddim.py file.

**Comment by beniz** [comment]
FYI, JoliGEN supports palette and has a working&tested implementation of DDIM here: https://github.com/jolibrain/joliGEN/blob/70e8ee47c92cab9ccd04a55a6c50e21a4b24004c/models/modules/diffusion_generator.py#L270 

This may help you integrate it here or use it directly via JG.

**Comment by AwayGithub** [comment]
@Saupatil07 @Janspiry
Hi! Here is my implementation of DDIM sampling in my image translation task, modified from JoliGEN code (Thanks to @beniz ). You can add into `network.py`.

    def restoration(
        self, 
        y_cond,
        sample_num,
        y_t=None, 
        y_0=None, 
        mask=None, 
        ddim_num_steps=300,
        ddim_eta=0,
        sampling_method='ddpm'
        ):
        
        if sampling_method == 'ddpm':
            return self.restoration_ddpm(y_cond, sample_num=sample_num, y_t=y_t, y_0=y_0, mask=mask)

        elif sampling_method == 'ddim':
            return self.restoration_ddim(y_cond, sample_num=sample_num, ddim_num_steps=ddim_num_steps, ddim_eta=ddim_eta, y_t=y_t, y_0=y_0, mask=mask)


    def restoration_ddim(self, y_cond, ddim_num_steps, ddim_eta, sample_num, y_t=None, y_0=None, mask=None):

        b, *_ = y_cond.shape
        assert ddim_num_steps > sample_num, 'ddim_num_steps must be greater than sample_num'
        sample_inter = (ddim_num_steps // sample_num)

        y_t = default(y_t, lambda: torch.randn_like(y_cond))
        ret_arr = y_t

        # linear step schedule
        tseq = list(
            np.linspace(0, self.num_timesteps - 1, ddim_num_steps).astype(int)
        )
        print("sample_num: ", sample_num)
        print("ddim_num_steps: ", ddim_num_steps)
        print("ddim_eta: ", ddim_eta)

        tlist = torch.zeros([y_t.shape[0]], device=y_t.device, dtype=torch.long)

        for i in tqdm(range(ddim_num_steps), desc='sampling loop time step', total=ddim_num_steps):
            tlist = tlist * 0 + tseq[-1 - i]

            if i != ddim_num_steps - 1:
                prevt = torch.ones_like(tlist, device=y_cond.device) * tseq[-2 - i]
            else:
                prevt = - torch.ones_like(tlist, device=y_cond.device)
            
            y_t = self.p_sample_ddim(y_t, tlist, prevt, ddim_num_steps, ddim_eta, y_cond=y_cond) 

            if mask is not None:
                y_t = y_0 * (1. - mask) + mask * y_t
            if (i + 1) % sample_inter == 0:
                ret_arr = torch.cat([ret_arr, y_t], dim=0)
            
        return y_t, ret_arr



    def p_sample_ddim(self, y_t, t, prevt, ddim_numsteps, ddim_eta, clip_denoised=True, y_cond=None):
        model_mean, model_log_variance = self.p_mean_variance_ddim(y_t=y_t, t=t, prevt=prevt, clip_denoised=clip_denoised, y_cond=y_cond, 
        ddim_numsteps=ddim_numsteps, ddim_eta=ddim_eta)
        noise = torch.randn_like(y_t) if any(t > 0) else torch.zeros_like(y_t)

        return model_mean + noise * (0.5 * model_log_variance).exp()



    def p_mean_variance_ddim(self, y_t, t, prevt, ddim_numsteps, ddim_eta, clip_denoised: bool, y_cond=None):
        noise_level = extract(self.gammas, t, x_shape=(1, 1)).to(y_t.device)
        noise_hat = self.denoise_fn(torch.cat([y_cond, y_t], dim=1), noise_level)
        y_0_hat = self.predict_start_from_noise(y_t, t=t, noise=noise_hat)
        if clip_denoised:
            y_0_hat.clamp_(-1., 1.)
        
        ## denoising formula in DDIM
        gamma_t = extract(self.gammas, t, y_t.shape)
        gamma_prev_t = extract(self.gammas_prev, prevt + 1, y_t.shape)

        sigma = ddim_eta * torch.sqrt((1. - gamma_prev_t) / (1. - gamma_t) * (1. - gamma_t / gamma_prev_t))
        p_var = sigma ** 2
        posterior_log_variance = torch.log(p_var)

        coef_eps = 1. - gamma_prev_t - p_var
        coef_eps[coef_eps < 0] = 0
        coef_eps = torch.sqrt(coef_eps)

        model_mean = (
            # torch.sqrt(gamma_prev_t) * (y_t - torch.sqrt(1.0 - gamma_t) * noise_hat) / torch.sqrt(gamma_t) + coef_eps * noise_hat
            torch.sqrt(gamma_prev_t) * y_0_hat + coef_eps * (y_t - torch.sqrt(gamma_t) * y_0_hat) / (torch.sqrt(1. - gamma_t))
        )
        
        return model_mean, posterior_log_variance



in my case, it works perfectly fine. Here are some of the results (output vs GT):

![Image](https://github.com/user-attachments/assets/191bedc5-0151-4921-b258-a0c5006046e2)
